---
title: "IceQream"
resource_files:
  - gastrulation-example/peak_intervals.rds
  - gastrulation-example/atac_scores.rds
  - gastrulation-example/additional_features.rds
  - gastrulation-example/gastrulation_intervals.tsv  
  - gastrulation-example/motif_energies.rds
  - model_report.pdf
  - mm10
---

```{r, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>"
)
```

## Introduction
IceQream (Interpretable Computational Engine for Quantitative Regression of Enhancer ATAC Motifs) is a package for regressing accessibility from DNA sequences. It models TF effective concentrations as latent variables that activate or repress regulatory elements in a nonlinear fashion, with possible contribution from pairwise interactions and synergistic chromosomal domain effects.

This vignette demonstrates how to use IceQream to analyze chromosome accessibility data, focusing on a sample of 5000 loci from a mouse gastrulation trajectory from Epiblast to Early nascent mesoderm. This analysis can help identify key regulatory elements and transcription factors involved in cellular differentiation processes, and understand quantitatively how they contribute to the observed changes in chromatin accessibility.

## Setup and Data Preparation

First, let's load the necessary packages and set up our environment:

```{r setup}
library(iceqream)
library(misha)
library(misha.ext)
```

### Creating a `misha` genome

We will start by creating a misha database for mm10 genome. If you already have an `mm10` misha database you can skip this part and load the genome using `gsetroot("path/to/mm10")`.

```{r create_genome, eval=FALSE}
gdb.create_genome("mm10")
```

```{r load_genome}
gsetroot("mm10")
```

### Downloading Example Data

See [below](#running-iceqream-on-your-own-data) for minimal instructions on how to use IceQream on your own data.

For this tutorial, we'll use pre-prepared data from a sample of the mouse gastrulation trajectory. Let's download and load this data:

```{r download_data, eval=FALSE}
download.file("https://iceqream.s3.eu-west-1.amazonaws.com/gastrulation-example.tar.gz", "gastrulation-example.tar.gz")
untar("gastrulation-example.tar.gz")
```

```{r load_data}
peak_intervals <- readr::read_rds("gastrulation-example/peak_intervals.rds")
atac_scores <- readr::read_rds("gastrulation-example/atac_scores.rds")
additional_features <- readr::read_rds("gastrulation-example/additional_features.rds")
normalization_intervals <- readr::read_tsv("gastrulation-example/gastrulation_intervals.tsv", show_col_types = FALSE)
```

Let's examine the structure of our input data:

```{r examine_data}
# Peak intervals
head(peak_intervals)
print(paste("Number of peaks:", nrow(peak_intervals)))

# ATAC scores
head(atac_scores)
print(paste("Number of peaks:", nrow(atac_scores)))
print(paste("Number of bins:", ncol(atac_scores)))

# Additional features
head(additional_features)
print(paste("Number of peaks:", nrow(additional_features)))
print(paste("Number of features:", ncol(additional_features)))
```

The `peak_intervals` dataframe contains the genomic positions of accessibility peaks. The `atac_scores` matrix contains ATAC-seq signal intensities for each peak across different stages of the trajectory. `additional_features` includes extra genomic features for each peak.

## Computing Motif Energies

The first step in the IceQream pipeline is to compute motif energies for each transcription factor model and each peak. This process can be computationally intensive, as it calculates energies for over 20,000 motifs from various databases.

```{r compute_motif_energies, eval=FALSE}
motif_energies <- compute_motif_energies(peak_intervals, motif_db, normalization_intervals = normalization_intervals)
```

However, for this tutorial, we're using pre-computed motif energies for a sample of 5000 loci, which are included in the example data:

```{r load_precomputed_energies}
motif_energies <- readr::read_rds("gastrulation-example/motif_energies.rds")
print(paste("Motif energy matrix dimensions:", paste(dim(motif_energies), collapse = " x ")))
```

This pre-computed matrix contains motif energies for our sample of peaks, allowing us to proceed with the analysis more quickly.

For a less memory and computationally intensive analysis on your own data, you can reduce the number of motifs used in the regression by taking a representative from the SCENIC clusters (1615) instead of all motifs (20,000+). This can be done by:

```{r reduce_motifs, eval=FALSE}
motif_energies <- compute_motif_energies(peak_intervals, motif_db_scenic_clusters, normalization_intervals = normalization_intervals)
```

## Running IceQream

Now we're ready to run the IceQream regression:

```{r run_iceqream}
traj_model <- iq_regression(
    peak_intervals = peak_intervals,
    atac_scores = atac_scores,
    motif_energies = motif_energies,
    normalize_energies = FALSE,
    additional_features = additional_features,
    norm_intervals = normalization_intervals,
    seed = 60427,
    frac_train = 0.8,
    max_motif_num = 30
    # include_interactions = TRUE # uncomment to include pairwise interactions
)
```

Let's examine the output:

```{r examine_output}
print(traj_model)
```

The `TrajectoryModel` object contains components such as the regression model, motif models, and predicted accessibility scores.

## Visualizing Results

Let's start with a scatter plot of observed vs. predicted accessibility changes:

```{r prediction_scatter, fig.width=7, fig.height=7}
plot_prediction_scatter(traj_model)
```

This plot shows how well our model predicts accessibility changes. Points closer to the diagonal line indicate better predictions. We measure the accuracy of the model using the coefficient of determination (R^2).

### Model report

Next, let's look at the model report, which provides detailed information about the motifs and their contributions:

```{r plot_report, eval = FALSE}
plot_traj_model_report(traj_model, filename = "model_report.pdf")
```

```{r show_report}
knitr::include_graphics("model_report.pdf")
```

### Interpreting the trajectory model report

The model report provides several key pieces of information (from left to right):

1. Motif logos show the inferred sequence preferences for each transcription factor model.
2. Response curves show how the accessibility changes as a function of binding energy for each TF.
3. Barplots show the coefficient of each non-linear term of every motif in the model.
4. Spatial model curves show the parameters of the spatial model for each TF. The RÂ² values indicate the predictive power each TF adds to the model.
5. Spatial curves show the frequency of each TF binding site along the peaks from the bottom 10% (blue) and top 10% (red) of the differential accessibility (dAP) distribution.
6. Boxplots show the distribution of ATAC differences (dAP, y-axis) for bins of binding energy (x-axis) for each TF.

## Renaming the motif models 

You can rename the motif models to more informative names, either manually using `rename_motif_models` or automatically using `match_traj_model_motif_names`:

```{r rename_motif_models}
names_map <- match_traj_model_motif_names(traj_model)
names_map
traj_model <- rename_motif_models(traj_model, names_map)
```

## Exporting the model

You can export the minimal model representation to a list of PBM in order to use infer its parameters on new data:

```{r export_model}
pbm_list <- traj_model_to_pbm_list(traj_model)
pbm_list
```

You can now use `pbm_list.compute` or `pbm_list.gextract`:

```{r infer_model}
new_intervals <- data.frame(
    chrom = rep("chr1", 3),
    start = c(3671720, 4412460, 4493400),
    end = c(3672020, 4412760, 4493700)
)
pbm_list.gextract(pbm_list, new_intervals)

# directly compute on sequences
seqs <- prego::intervals_to_seq(new_intervals)
seqs
pbm_list.compute(pbm_list, seqs)
```

## Running IceQream on your own data

To run IceQream on your own data, you will need to provide the following inputs:

1. Genomic positions of peaks (as a dataframe with columns `chrom`, `start`, `end`, `peak_name`), optionally it can have a `const` column indicating constitutive loci.
2. ATAC scores (as a matrix with rows corresponding to peaks and columns corresponding to bins).
3. (Optional) Additional features (as a data frame with rows corresponding to peaks and columns corresponding to features).

You can then follow the steps outlined in this vignette to compute motif energies, run the regression and analyze the results.
